.sandbox/provision/lifecycle-asg/common.sh

```diff
@@ -0,0 +1,101 @@
+function claim_instance_from_asg() {
+    # check is there any running instance already
+    local instance_with_sandbox_id=""
+    instance_with_sandbox_id="$(aws ec2 describe-instances --filters Name=tag:$SANDBOX_ID_TAG,Values="$SANDBOX_ID" Name=instance-state-name,Values=running)"
+    if [[ "$(jq -cMr '.Reservations[0].Instances | length' <<< "$instance_with_sandbox_id")" -eq 0 ]]; then 
+        local instance_id=""
+        instance_id="$(aws autoscaling describe-auto-scaling-groups --auto-scaling-group-name "$ASG_NAME" --no-paginate --query AutoScalingGroups[0].Instances[0].InstanceId --output text)"
+
+        # detach-instances is an atomic operation and a single instance can not be deatched twice. 
+        # This mechanism is used to prevent race condition between two or more claims that happening at the same time. The second claim would fail.
+        aws autoscaling detach-instances --instance-ids "$instance_id" --auto-scaling-group-name "$ASG_NAME" --no-should-decrement-desired-capacity >/dev/null
+
+        # Two tags are created for the detached instance
+        # - Sandbox, this is for referencing purpose and should not be used as an ID during looking up.
+        # - SandboxID, this is the actual identification of the EC2 instance.
+        # An EC2 instance is deemed as leaked if:
+        # - SandboxID tag is missing
+        # - Some ASG flagging tag exists, e.g. sandbox-asg=true
+        # - The instance is detached from ASG.
+        aws ec2 create-tags --resources "$instance_id" --tags Key=$SANDBOX_NAME_TAG,Value="$SANDBOX_NAME" --tags Key=$SANDBOX_ID_TAG,Value="$SANDBOX_ID"
+
+        echo "$instance_id"
+    else 
+        jq -cMr '.Reservations[0].Instances[0].InstanceId' <<< "$instance_with_sandbox_id"
+    fi
+}
+
+function attach_volume_if_needed() {
+    local volume_id="$1"
+    local instance_id="$2"
+    local volume=""
+
+    aws ec2 wait volume-available --volume-ids "$volume_id"
+    volume="$(aws ec2 describe-volumes --volume-id "$volume_id")"
+    [[ "$(jq -cMr ".Volumes[0].Attachments | length" <<< "$volume")" -gt 0 ]] || {
+        aws ec2 attach-volume --volume-id "$volume_id" --instance-id "$instance_id" --device "$DEVICE_NAME"
+    }
+}
+
+# retrieve_password INSTANCE_ID KEY_FILE
+function retrieve_password() {
+    local instance_id="$1"
+    local ec2_ssh_key_file="$2"
+    local password_data=""
+
+    password_data="$(aws ec2 get-password-data --instance-id "$instance_id" --query 'PasswordData' --output text)"
+
+    local retry_count=0
+    while [[ "$retry_count" -lt "$MAX_RETRIES" ]]; do
+        password_data="$(aws ec2 get-password-data --instance-id "$instance_id" --query 'PasswordData' --output text)"
+        [[ -z "$password_data" ]] || break
+        retry_count=$((retry_count + 1))
+        sleep 10
+    done
+
+    echo "$password_data" | base64 -d  | openssl pkeyutl  -decrypt -inkey "$ec2_ssh_key_file"
```

Comment:
What about the case while loop is exhausted? (`password_data` is empty)

Reply:
I think in this case, the best operation is to re-create the sandbox or restart the sandbox. 

I came across this once and it normally indicates the generated EC2 instance has some problem and there is no way to fix it.

Reply:
I mean from the script perspective, the logic must cover all possible error branches properly, we can't leave some cases unhandled, regardless what it will result in.